File: backend/app/routes/places.py
"""장소 검색 프록시 — 카카오 로컬 API 키를 클라이언트에 노출하지 않기 위한 백엔드 프록시.

현재 제공: 노인 돌봄센터 키워드 검색(GET /places/care-centers).
- 카카오 키워드 검색(/v2/local/search/keyword.json)을 여러 돌봄 키워드로 병렬 호출 후 병합한다.
- category_name 에 '사회복지' 가 포함된 항목만 남긴다(키워드 오검색 제거).
- place id(또는 위경도) 기준으로 중복을 제거하고 거리순으로 정렬한다.
- 카카오 에러/타임아웃 시 예외를 내지 않고 빈 배열 + 에러 코드를 반환한다(프론트 빈 상태 처리).

응답 항목은 /hospital/search-nearby 와 동일한 필드(name·address·distance·phone·url 등)를
사용해 프론트에서 같은 장소 타입을 재사용할 수 있게 한다(여기에 lat·lng 추가 노출).
"""

import asyncio
import os

import httpx
from fastapi import APIRouter, Query
from pydantic import BaseModel

KAKAO_REST_KEY = os.getenv("KAKAO_REST_KEY", "c19535cd0cd962a7ddbc759cec982d36")
KAKAO_KEYWORD_URL = "https://dapi.kakao.com/v2/local/search/keyword.json"

# 노인 돌봄센터 키워드 — 카카오 키워드 검색에 병렬 질의 후 병합한다.
CARE_CENTER_KEYWORDS = ["재가복지센터", "주야간보호센터", "노인복지관", "노인요양센터"]
# 키워드 오검색을 거르기 위한 카테고리 필터 토큰(예: "사회,공공기관 > 사회복지시설 > ...").
SOCIAL_WELFARE_TOKEN = "사회복지"

router = APIRouter(prefix="/places", tags=["places"])


class Place(BaseModel):
    id: str | None = None
    name: str
    address: str
    lat: float | None = None
    lng: float | None = None
    distance: str | None = None
    phone: str | None = None
    url: str | None = None


class CareCentersResponse(BaseModel):
    items: list[Place]
    # 정상(0건 포함) 시 None. 카카오 통신 실패 시 코드 문자열.
    error: str | None = None


async def _search_keyword(
    client: httpx.AsyncClient, keyword: str, lat: float, lng: float, radius: int
) -> list[dict]:
    params = {
        "query": keyword,
        "x": str(lng),
        "y": str(lat),
        "radius": str(radius),
        "sort": "distance",
    }
    headers = {"Authorization": f"KakaoAK {KAKAO_REST_KEY}"}
    res = await client.get(KAKAO_KEYWORD_URL, headers=headers, params=params, timeout=10)
    res.raise_for_status()
    return res.json().get("documents", [])


@router.get("/care-centers", response_model=CareCentersResponse)
async def search_care_centers(
    lat: float = Query(..., description="위도"),
    lng: float = Query(..., description="경도"),
    radius: int = Query(2000, description="반경(m)"),
):
    """위경도 기준 주변 노인 돌봄센터를 검색한다 (카카오 키워드 검색 프록시).

    여러 키워드 병렬 질의 → category_name 에 '사회복지' 포함만 필터 → 중복 제거 → 거리순 정렬.
    카카오 에러/타임아웃 시 빈 배열 + 에러 코드를 반환한다(예외를 던지지 않음).
    """
    try:
        async with httpx.AsyncClient() as client:
            results = await asyncio.gather(
                *[_search_keyword(client, kw, lat, lng, radius) for kw in CARE_CENTER_KEYWORDS]
            )
    except httpx.TimeoutException:
        return CareCentersResponse(items=[], error="KAKAO_TIMEOUT")
    except Exception:
        return CareCentersResponse(items=[], error="KAKAO_ERROR")

    merged: dict[str, Place] = {}
    for docs in results:
        for doc in docs:
            category = doc.get("category_name") or ""
            if SOCIAL_WELFARE_TOKEN not in category:
                continue
            # 중복 제거: place id 우선, 없으면 위경도 조합으로 식별.
            place_id = doc.get("id") or f"{doc.get('x')},{doc.get('y')}"
            if place_id in merged:
                continue
            x, y = doc.get("x"), doc.get("y")
            merged[place_id] = Place(
                id=doc.get("id"),
                name=doc.get("place_name") or "",
                address=doc.get("road_address_name") or doc.get("address_name") or "",
                lat=float(y) if y else None,
                lng=float(x) if x else None,
                distance=doc.get("distance"),
                phone=doc.get("phone") or None,
                url=doc.get("place_url") or None,
            )

    items = sorted(
        merged.values(),
        key=lambda p: int(p.distance) if p.distance and p.distance.isdigit() else 10**9,
    )
    return CareCentersResponse(items=items, error=None)

File: frontend/src/api/places.ts
// 장소 검색 API — 카카오는 프론트에서 직접 호출하지 않고 백엔드 프록시(/places, /hospital/geocode)만 호출.
// mock-first: EXPO_PUBLIC_AUTH_API_MODE 가 "real" 일 때만 백엔드 호출, 그 외에는 더미 데이터.
import { getToken } from "./session";

const API_BASE_URL = (process.env.EXPO_PUBLIC_API_BASE_URL ?? "http://localhost:8000").replace(
  /\/$/,
  "",
);
const IS_REAL = process.env.EXPO_PUBLIC_AUTH_API_MODE === "real";

// /places/care-centers 및 /hospital/search-nearby 와 동일한 장소 형태(타입 재사용).
export interface Place {
  id?: string;
  name: string;
  address: string;
  lat?: number;
  lng?: number;
  distance?: string; // 카카오 원본 미터 문자열. 표시 포맷은 화면에서 처리.
  phone?: string;
  url?: string;
}

export interface CareCentersResult {
  items: Place[];
  error: string | null; // null=정상(0건 포함). 그 외=통신 실패 코드.
}

export type Coords = { lat: number; lng: number };

// mock 모드에서 보여줄 더미 돌봄센터 2~3건.
const MOCK_CARE_CENTERS: Place[] = [
  { id: "mock-1", name: "햇살재가복지센터", address: "광주 서구 상무대로 123", distance: "420", phone: "062-000-0001", lat: 35.152, lng: 126.852 },
  { id: "mock-2", name: "어울림주야간보호센터", address: "광주 서구 운천로 45", distance: "880", phone: "062-000-0002", lat: 35.149, lng: 126.857 },
  { id: "mock-3", name: "한마음노인복지관", address: "광주 서구 계수로 51", distance: "1240", phone: "062-000-0003", lat: 35.160, lng: 126.856 },
];

// 주소 → 좌표 (백엔드 카카오 지오코딩 프록시 재사용). real 모드 전용.
export async function geocodeAddress(address: string): Promise<Coords | null> {
  const token = await getToken();
  const res = await fetch(`${API_BASE_URL}/hospital/geocode?address=${encodeURIComponent(address)}`, {
    headers: token ? { Authorization: `Bearer ${token}` } : {},
  });
  if (!res.ok) return null;
  const json = (await res.json()) as { lat: number; lng: number };
  return { lat: json.lat, lng: json.lng };
}

// 좌표 기준 돌봄센터 검색 (백엔드 /places/care-centers 프록시). mock 모드는 더미 반환.
export async function searchCareCenters(
  coords: Coords | null,
  radius = 2000,
): Promise<CareCentersResult> {
  if (!IS_REAL) {
    return { items: MOCK_CARE_CENTERS, error: null };
  }
  if (!coords) return { items: [], error: null };
  try {
    const token = await getToken();
    const res = await fetch(
      `${API_BASE_URL}/places/care-centers?lat=${coords.lat}&lng=${coords.lng}&radius=${radius}`,
      { headers: token ? { Authorization: `Bearer ${token}` } : {} },
    );
    if (!res.ok) return { items: [], error: "REQUEST_FAILED" };
    return (await res.json()) as CareCentersResult;
  } catch {
    return { items: [], error: "NETWORK_ERROR" };
  }
}
